[ONNX] Export nn.Module call as ONNX local function (#63589)

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66149 [ONNX] Update slice process shape to support rank only inference (#65782)
* #66147 [ONNX] Symbolic: dynamic input for OneHot, bool for Einsum (#65940)
* #66146 [ONNX] Modify softplus symbolic to support beta!=1 (#65001)
* #66145 [ONNX] Deprecate fold_if pass (#65697)
* #66144 [ONNX] Add diagonal symbolic (#64454)
* #66143 ONNX: Delete or document skipped ORT tests (#64470)
* #66142 [ONNX] Add warning for inplace updates on tensor.shape in tracing mode (#63170)
* #66141 [ONNX] Update sum symbolic to handle dtypes (#64289)
* **#66140 [ONNX] Export nn.Module call as ONNX local function (#63589)**

* Add new argument to export api to enable users specifying `nn.Module` classes that they wish to be exported as local function in ONNX model.
* Refactor `torch/csrc/jit/serialization/export.cpp`, and remove redundant `EncoderBase` class.
* ~~Contains changes from #63268~~
* Depends on #63716 to update onnx submodule.

Co-authored-by: BowenBao <bowbao@microsoft.com>

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D31424098](https://our.internmc.facebook.com/intern/diff/D31424098)

BowenBao committed:
[ONNX] Export nn.Module call as ONNX local function (#63589)

* Add new argument to export api to enable users specifying `nn.Module` classes that they wish to be exported as local function in ONNX model.
* Refactor `torch/csrc/jit/serialization/export.cpp`, and remove redundant `EncoderBase` class.
* ~~Contains changes from #63268~~
* Depends on #63716 to update onnx submodule.

Co-authored-by: BowenBao <bowbao@microsoft.com>

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export nn.Module call as ONNX local function (#63589)"


* Add new argument to export api to enable users specifying `nn.Module` classes that they wish to be exported as local function in ONNX model.
* Refactor `torch/csrc/jit/serialization/export.cpp`, and remove redundant `EncoderBase` class.
* ~~Contains changes from #63268~~
* Depends on #63716 to update onnx submodule.

Co-authored-by: BowenBao <bowbao@microsoft.com>

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/).

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31424098).

ezyang commented:
we probably have to do a tp2 upgrade of onnx in that case

houseroad commented:
@BowenBao could you try rebase on top of #66427 and see whether the build issue is gone?

BowenBao committed:
Update on "[ONNX] Export nn.Module call as ONNX local function (#63589)"


* Add new argument to export api to enable users specifying `nn.Module` classes that they wish to be exported as local function in ONNX model.
* Refactor `torch/csrc/jit/serialization/export.cpp`, and remove redundant `EncoderBase` class.
* ~~Contains changes from #63268~~
* Depends on #63716 to update onnx submodule.

Co-authored-by: BowenBao <bowbao@microsoft.com>

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D31424098](https://our.internmc.facebook.com/intern/diff/D31424098)

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export nn.Module call as ONNX local function (#63589)"


* Add new argument to export api to enable users specifying `nn.Module` classes that they wish to be exported as local function in ONNX model.
* Refactor `torch/csrc/jit/serialization/export.cpp`, and remove redundant `EncoderBase` class.
* ~~Contains changes from #63268~~
* Depends on #63716 to update onnx submodule.

Co-authored-by: BowenBao <bowbao@microsoft.com>

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D31424098](https://our.internmc.facebook.com/intern/diff/D31424098)

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export nn.Module call as ONNX local function (#63589)"


* Add new argument to export api to enable users specifying `nn.Module` classes that they wish to be exported as local function in ONNX model.
* Refactor `torch/csrc/jit/serialization/export.cpp`, and remove redundant `EncoderBase` class.
* ~~Contains changes from #63268~~
* Depends on #63716 to update onnx submodule.

Co-authored-by: BowenBao <bowbao@microsoft.com>

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D31424098](https://our.internmc.facebook.com/intern/diff/D31424098)

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export nn.Module call as ONNX local function (#63589)"


* Add new argument to export api to enable users specifying `nn.Module` classes that they wish to be exported as local function in ONNX model.
* Refactor `torch/csrc/jit/serialization/export.cpp`, and remove redundant `EncoderBase` class.
* ~~Contains changes from #63268~~
* Depends on #63716 to update onnx submodule.

Co-authored-by: BowenBao <bowbao@microsoft.com>

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D31424098](https://our.internmc.facebook.com/intern/diff/D31424098)

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Export nn.Module call as ONNX local function (#63589)"


* Add new argument to export api to enable users specifying `nn.Module` classes that they wish to be exported as local function in ONNX model.
* Refactor `torch/csrc/jit/serialization/export.cpp`, and remove redundant `EncoderBase` class.
* ~~Contains changes from #63268~~
* Depends on #63716 to update onnx submodule.

Co-authored-by: BowenBao <bowbao@microsoft.com>

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D31424098](https://our.internmc.facebook.com/intern/diff/D31424098)

[ghstack-poisoned]
